nextstrain/ncov-ingest
# Add nextstrain clades to metadata (version for running locally)

### Description of proposed changes    

During GISAID ingest, it is proposed to produce a version of `metadata.tsv` file which contains an additional column "Nextstrain_clade", containing clade deduced using Nextclade CLI tool.

Thus, the "nextmeta" published on GISAID will contain not only Pangolin lineages and GISAID clades, but also Nextstrain clades.

---

<sub>
Work in progress: Currently only has implementation for local runs.
</sub>

---

<sub>
TODO: run for a few days in addition to the current vanilla ingest and compare the results. Port to GitHub-action-based script.
</sub>

---

In order to make this happen, this PR introduces 2 distinct tasks:

##### One-time preparation task ("full run"):

Ensures that all currently existing GISAID sequences (~100k) are processed and have a clade assigned. It's too compute-intensive for GitHub actions and must be done locally, on premises or on a high-performance cloud instance. Runs in a few hours on a laptop, in background or overnight.

The same task can be used for periodic mass-synchronization of clades, to account for metadata and sequence changes on GISAID.

Implementation details:
 - have fresh `sequences.fasta`, `metadata.tsv`
 - split `sequences.fasta` into batches: `sequences.fasta.batch-xxxxx.tsv`
 - for each batch (in parallel): run `nextclade.js`, get `sequences.fasta.batch-xxxxx.tsv`, containing clades
 - concatenate results into `nextclade.clades.tsv`
 - join `metadata.tsv` and `nextclade.clades.tsv`, such that `metadata2.tsv` has an additional column "Nextstrain_clade"



##### Daily task:

Runs as a part of daily ingest. Takes only sequences that have no clades, assigns clades, merges them into the new metadata. Should take no more than a minute or a few for a usual daily batch of sequences.

Implementation details:

 - during ingest, have fresh `metadata.tsv`, `sequences.fasta`, and old `nextclade.clades.tsv`
 - filter `sequences.fasta`: get a list of sequences that are in `sequences.fasta`, but are not in `nextclade.clades.tsv`, save them to `nextclade.sequences.fasta`
 - run nextclade on `nextclade.sequences.fasta`, get `nextclade.clades.new.tsv`
 - add `nextclade.clades.new.tsv` to `nextclade.clades.tsv`
 - join `metadata.tsv` and `nextclade.clades.tsv`
 - TODO: the final implementation for GitHub actions should persist `nextclade.clades.tsv` on S3


##### Additional considerations and decisions:

 - Name of the new column. Currently "Nextstrain_clade".
 
 - How and how often to run "full runs".

 - Whether we need to implement matching by GISAID ID. Currently matching of clade to metadata is done by sequence name. These names might change on GISAID. A more reliable mapping can be implemented by using GISAID-specific stable identifier. Nextclade has no access to metadata and only operates on sequence names. So this might not be currently possible to do reliably without processing sequences one by one, which might be very slow.
  Additionally, matching by name may produce inaccurate results if names are not unique and sequences with the same name are supposed to be in different clades.

 - Clades is only a tiny portion of Nextclade's output. We may consider more things to be saved, either for internal use or for publishing.


### Related issue(s)  
N/A

### Testing

This is based on local ingest branch (see #89)

 - setup local ingest environment as described in #89 

 - the script currently assumes that the `master` version of `nextclade.js` is available globally. That's the only version that has the new `--output-tsv-clades-only` flag. You can build it as described in [README](https://github.com/nextstrain/nextclade/blob/master/packages/cli/README.md#build) and it appears in packages/cli/dist. Then I just symlinked it to one of my directories in `PATH` (`~/bin`).
   It will be available on npm after we finalize this business here, so that simple `npm install -g @neherlab/nextclade` will do then.

 - for "full run":

    ```bash
    nice -19 ionice -c3 ./bin/gisaid-get-all-clades
    ```
    (scales with the number of available threads)

 - observe `data/gisaid/outputs/metadata2.tsv` which contains clades for all sequences (excluding the ones that failed to align during nextclade step)

 - observe `data/gisaid/outputs/nextclade.clades.tsv` with clade mapping (and no other metadata)

 - move `data/gisaid/outputs/nextclade.clades.tsv` to `data/gisaid/inputs/nextclade.clades.tsv`
   (simulating the transition to the next day - inputs of the today are the outputs of yesterday)

 - for daily ingest:
   run as described in #89 (the new steps are added to the `--ingest` part)

 - observe `data/gisaid/outputs/metadata2.tsv` which contains clades for all sequences, including today's

### Thank you for contributing to Nextstrain!

🙏


@rneher @emmahodcroft Could you please expand on motivation and other considerations for this feature?
